[Gradient Compression] Add CppCommHook subclass for supporting the C++ API of communication hook.

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #46959 [Gradient Compression] Surface C++ comm hooks to Python API as built-in comm hooks
* #46701 [Gradient Compression] Provide 2 default C++ comm hooks
* **#46566 [Gradient Compression] Add CppCommHook subclass for supporting the C++ API of communication hook.**

Only provides an interface. Some built-in implementations will be provided in a follow-up commit.

riginal PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24379460](https://our.internmc.facebook.com/intern/diff/D24379460/)

Commit:
[Gradient Compression] Add CppCommHook subclass for supporting the C++ API of communication hook.

Only provides an interface. Some built-in implementations will be provided in a follow-up commit.

riginal PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24379460](https://our.internmc.facebook.com/intern/diff/D24379460/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Add CppCommHook subclass for supporting the C++ API of communication hook."


Only provides an interface. Some built-in implementations will be provided in a follow-up commit.

riginal PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24379460](https://our.internmc.facebook.com/intern/diff/D24379460/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Add CppCommHook subclass for supporting the C++ API of communication hook."


Only provides an interface. Some built-in implementations will be provided in a follow-up commit.

riginal PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24379460](https://our.internmc.facebook.com/intern/diff/D24379460/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Add CppCommHook subclass for supporting the C++ API of communication hook."


Only provides an interface. Some built-in implementations will be provided in a follow-up commit.

riginal PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24379460](https://our.internmc.facebook.com/intern/diff/D24379460/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/c10d/comm.h:
```diff
@@ -45,7 +45,7 @@ class TORCH_API CommHookInterface {
   // asynchronously, Returns a future that holds the communication results.
   virtual c10::intrusive_ptr<torch::jit::Future> runHook(
-      const GradBucket& bucket) = 0;
+      GradBucket& bucket) = 0;
```
What is the reason to remove const from here?

SciPioneer replied:
It's because of the subsequent change, which updates the bucket in place.

pritamdamania87 commented on torch/csrc/distributed/c10d/comm.h:
```diff
@@ -67,14 +67,35 @@ class TORCH_PYTHON_API PythonCommHook : public CommHookInterface {
+  // This can be a more generic state if needed.
+  // Note that std::optional<ProcessGroup> cannot be used, since ProcessGroup is
+  // an abstract class.
+  ProcessGroup* process_group_; // Not owned.
```
Why is this not owned, we can just pass the std::shared_ptr from c10d/reducer here? It's usually just safer to have a shared_ptr here.

SciPioneer replied:
Will the lifecycle of this comm hook outlive the reducer that initially owns this process group? I don't think so as comm_hook_ (that keeps this field) is a member of reducer. Therefore, a raw pointer should be sufficient here.

I used a raw pointer mainly because its assignment is cheaper than shared pointer (i.e., no increment of 'atomic' reference counter). As long as a data member does not need to be owned, I generally do not prefer a smart pointer wrapper.

Commit:
Update on "[Gradient Compression] Add CppCommHook subclass for supporting the C++ API of communication hook."


Only provides an interface. Some built-in implementations will be provided in a follow-up commit.

riginal PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24379460](https://our.internmc.facebook.com/intern/diff/D24379460/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/c10d/comm.h:
```diff
@@ -68,16 +68,39 @@ class TORCH_PYTHON_API PythonCommHook : public CommHookInterface {
+ public:
+  explicit CppCommHook(
+      std::function<c10::intrusive_ptr<
+          torch::jit::Future>(ProcessGroup*, GradBucket&)>& hook,
```
Another option here might be to make this class a template class and do this:

```
template <typename T>
class TORCH_API CppCommHook : public CommHookInterface {
  public:
    explicit CppCommHook(
      std::function<c10::intrusive_ptr<
          torch::jit::Future>(T, GradBucket&)>& hook,
```

This way users can use whatever object they want to as their state, for example:

```class FP16Compress : public CommHookInterface<std::shared_ptr<ProcessGroup>> {}```

SciPioneer replied:
Considered template class before. One minor concern is that, if the state is null, there isn't a nice way to represent this with T, as T can be anything.

I chose ProcessGroup* mainly because I have only seen the cases that the state here is either a process group or null. On the other hand, I cannot use std::optional<ProcessGroup>, optional cannot be applied to an abstract class.

pritamdamania87 replied:
We can pass `nullptr` for T in case the state is null? And the subclass could be defined as:

`class MyCompress : public CommHookInterface<void*> {}`

SciPioneer replied:
Fair enough. Let's use void* to represent a null state.

Commit:
Update on "[Gradient Compression] Add CppCommHook subclass for supporting the C++ API of communication hook."


Only provides an interface. Some built-in implementations will be provided in a follow-up commit.

riginal PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24379460](https://our.internmc.facebook.com/intern/diff/D24379460/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Add CppCommHook subclass for supporting the C++ API of communication hook."


Only provides an interface. Some built-in implementations will be provided in a follow-up commit.

riginal PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24379460](https://our.internmc.facebook.com/intern/diff/D24379460/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Add CppCommHook subclass for supporting the C++ API of communication hook."


Only provides an interface. Some built-in implementations will be provided in a follow-up commit.

riginal PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24379460](https://our.internmc.facebook.com/intern/diff/D24379460/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/c10d/comm.h:
```diff
@@ -68,16 +68,39 @@ class TORCH_PYTHON_API PythonCommHook : public CommHookInterface {
+template <typename T>
+class TORCH_API CppCommHookInterface : public CommHookInterface {
+ public:
+  explicit CppCommHookInterface(T* state = nullptr) : state_(state) {}
```
Why is this `T* state`, it should just be `T state`. With T* the interface is not as flexible since users might not want to pass a naked pointer here in some cases. You can always pass `ProcessGroup*` as an argument even with `T state`.

pritamdamania87 commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -700,7 +700,8 @@ void Reducer::mark_bucket_ready(size_t bucket_index) {
       bucket.work = process_group_->allreduce(tensors);
     } else {
-      bucket.future_work = comm_hook_->runHook(GradBucket(tensors));
+      GradBucket grad_bucket(tensors);
+      bucket.future_work = comm_hook_->runHook(grad_bucket);
```
We can consolidate this into just one line as it was previously?

Commit:
Update on "[Gradient Compression] Add CppCommHook subclass for supporting the C++ API of communication hook."


Only provides an interface. Some built-in implementations will be provided in a follow-up commit.

riginal PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24379460](https://our.internmc.facebook.com/intern/diff/D24379460/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Add CppCommHook subclass for supporting the C++ API of communication hook."


Only provides an interface. Some built-in implementations will be provided in a follow-up commit.

riginal PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24379460](https://our.internmc.facebook.com/intern/diff/D24379460/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Add CppCommHook subclass for supporting the C++ API of communication hook."


Only provides an interface. Some built-in implementations will be provided in a follow-up commit.

riginal PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24379460](https://our.internmc.facebook.com/intern/diff/D24379460/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Add CppCommHook subclass for supporting the C++ API of communication hook."


Only provides an interface. Some built-in implementations will be provided in a follow-up commit.

riginal PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24379460](https://our.internmc.facebook.com/intern/diff/D24379460/)

[ghstack-poisoned]
